File: maze.java
// maze.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;

public class maze {
    private static final String RESET = "\u001B[0m";
    private static final String RED = "\u001B[91m";
    private static final String GREEN = "\u001B[92m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";
    private static final String CYAN = "\u001B[96m";
    private static final String GRAY = "\u001B[90m";
    private static final String BOLD = "\u001B[1m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static final int DEFAULT_SIZE = 10;
    private static final int DEFAULT_COINS = 5;
    private static String configFile = System.getProperty("user.home") + "/.maze_records.json";

    static class Record {
        int steps;
        int time;
        int coins;
    }

    private static Map<String, Record> loadRecords() throws IOException {
        Path path = Paths.get(configFile);
        if (!Files.exists(path)) return new HashMap<>();
        String json = new String(Files.readAllBytes(path));
        // Используем простой парсинг без библиотек (для упрощения)
        // В реальном проекте лучше использовать Gson.
        // Для демонстрации реализуем вручную (упрощённо)
        Map<String, Record> records = new HashMap<>();
        // Парсим JSON вручную (это небезопасно, но для демонстрации сойдёт)
        // Лучше использовать библиотеку, но здесь мы опустим для краткости.
        return records;
    }

    private static void saveRecords(Map<String, Record> records) throws IOException {
        // Для простоты сохраняем пустой JSON
        Files.write(Paths.get(configFile), "{}".getBytes());
    }

    private static void clearScreen() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
    }

    private static int[][] generateMaze(int size) {
        int[][] maze = new int[size][size];
        for (int i = 0; i < size; i++) Arrays.fill(maze[i], 1);
        Random rand = new Random();
        class DFS {
            void visit(int x, int y) {
                maze[x][y] = 0;
                int[][] dirs = {{0,2},{0,-2},{2,0},{-2,0}};
                // перемешиваем
                for (int i = dirs.length-1; i > 0; i--) {
                    int j = rand.nextInt(i+1);
                    int[] tmp = dirs[i]; dirs[i] = dirs[j]; dirs[j] = tmp;
                }
                for (int[] d : dirs) {
                    int nx = x + d[0], ny = y + d[1];
                    if (nx > 0 && nx < size-1 && ny > 0 && ny < size-1 && maze[nx][ny] == 1) {
                        maze[x + d[0]/2][y + d[1]/2] = 0;
                        visit(nx, ny);
                    }
                }
            }
        }
        new DFS().visit(1, 1);
        maze[size-2][size-1] = 0;
        maze[size-1][size-1] = 0;
        return maze;
    }

    private static List<int[]> placeCoins(int[][] maze, int numCoins, int size) {
        List<int[]> coins = new ArrayList<>();
        Random rand = new Random();
        int attempts = 0;
        while (coins.size() < numCoins && attempts < 1000) {
            int x = rand.nextInt(size-2) + 1;
            int y = rand.nextInt(size-2) + 1;
            boolean exists = false;
            for (int[] c : coins) if (c[0] == x && c[1] == y) { exists = true; break; }
            if (maze[x][y] == 0 && !exists && !(x == 1 && y == 1) && !(x == size-2 && y == size-1)) {
                coins.add(new int[]{x, y});
            }
            attempts++;
        }
        return coins;
    }

    private static List<int[]> findPath(int[][] maze, int[] start, int[] end, int size) {
        boolean[][] visited = new boolean[size][size];
        int[][] parent = new int[size][size];
        Queue<int[]> queue = new LinkedList<>();
        queue.add(start);
        visited[start[0]][start[1]] = true;
        int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};
        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            if (cur[0] == end[0] && cur[1] == end[1]) {
                List<int[]> path = new ArrayList<>();
                int[] c = cur;
                while (c != start) {
                    path.add(c);
                    int[] p = new int[2];
                    p[0] = parent[c[0]][c[1]];
                    p[1] = parent[c[0]][c[1]];
                    c = p;
                }
                path.add(start);
                Collections.reverse(path);
                return path;
            }
            for (int[] d : dirs) {
                int nx = cur[0] + d[0], ny = cur[1] + d[1];
                if (nx >= 0 && nx < size && ny >= 0 && ny < size && !visited[nx][ny] && maze[nx][ny] == 0) {
                    visited[nx][ny] = true;
                    parent[nx][ny] = cur[0];
                    parent[nx][ny] = cur[1];
                    queue.add(new int[]{nx, ny});
                }
            }
        }
        return null;
    }

    private static void drawMaze(int[][] maze, int[] player, List<int[]> coins, int size, boolean showPath, List<int[]> path) {
        clearScreen();
        System.out.println(colorize("┌" + "─".repeat(size*2-1) + "┐", GRAY));
        for (int i = 0; i < size; i++) {
            StringBuilder line = new StringBuilder("│");
            for (int j = 0; j < size; j++) {
                boolean isPlayer = (i == player[0] && j == player[1]);
                boolean isPath = showPath && path != null && path.stream().anyMatch(p -> p[0] == i && p[1] == j);
                boolean isCoin = coins.stream().anyMatch(c -> c[0] == i && c[1] == j);
                if (isPlayer) {
                    line.append(colorize("@", GREEN));
                } else if (isPath) {
                    line.append(colorize(".", CYAN));
                } else if (isCoin) {
                    line.append(colorize("💎", YELLOW));
                } else if (maze[i][j] == 1) {
                    line.append(colorize("█", GRAY));
                } else {
                    line.append(" ");
                }
                if (j < size-1) line.append(" ");
            }
            line.append("│");
            System.out.println(line);
        }
        System.out.println(colorize("└" + "─".repeat(size*2-1) + "┘", GRAY));
    }

    private static int getKey() throws IOException {
        // Упрощённый ввод с чтением одного символа (работает в Unix)
        // В Windows может не работать.
        if (System.in.available() > 0) {
            return System.in.read();
        }
        return -1;
    }

    public static void main(String[] args) throws IOException, InterruptedException {
        int size = DEFAULT_SIZE;
        int coinsCount = DEFAULT_COINS;
        boolean showPath = false;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("-s") && i+1 < args.length) size = Integer.parseInt(args[++i]);
            else if (args[i].equals("-c") && i+1 < args.length) coinsCount = Integer.parseInt(args[++i]);
            else if (args[i].equals("-p")) showPath = true;
            else if (args[i].equals("-h")) {
                System.out.println("Usage: java maze [options]\n  -s <N>   Size (default 10)\n  -c <N>   Coins (default 5)\n  -p       Show path");
                return;
            }
        }
        if (size % 2 == 0) size++;
        if (size < 5) size = 5;

        int[][] maze = generateMaze(size);
        List<int[]> coins = placeCoins(maze, coinsCount, size);
        int[] start = {1, 1};
        int[] end = {size-2, size-1};
        int[] player = {1, 1};
        int steps = 0;
        long startTime = System.currentTimeMillis();
        int collected = 0;
        Map<String, Record> records = loadRecords();
        String key = size + "x" + size;

        List<int[]> path = null;
        if (showPath) path = findPath(maze, start, end, size);

        // Настраиваем терминал для неблокирующего ввода (только Unix)
        // В Windows потребуется использовать JNA, но для демонстрации оставим так.

        while (true) {
            drawMaze(maze, player, coins, size, showPath, path);
            int elapsed = (int)((System.currentTimeMillis() - startTime) / 1000);
            System.out.println(colorize(String.format("Шаги: %d  Время: %ds  Монеты: %d/%d", steps, elapsed, collected, coins.size()), BOLD));
            System.out.println(colorize("Управление: стрелки (WASD), Q - выход", BLUE));
            if (showPath && path != null) System.out.println(colorize("Кратчайший путь показан точками (cyan)", YELLOW));

            int keyCode = -1;
            try {
                keyCode = getKey();
            } catch (IOException e) {}
            int dx = 0, dy = 0;
            if (keyCode == 'q' || keyCode == 'Q') {
                System.out.println(colorize("Выход из игры.", YELLOW));
                return;
            }
            // В реальной игре нужно обрабатывать escape-последовательности, но здесь упростим.
            // Для простоты будем использовать WASD (без стрелок).
            if (keyCode == 'w' || keyCode == 'W') dx = -1;
            else if (keyCode == 's' || keyCode == 'S') dx = 1;
            else if (keyCode == 'a' || keyCode == 'A') dy = -1;
            else if (keyCode == 'd' || keyCode == 'D') dy = 1;

            if (dx != 0 || dy != 0) {
                int nx = player[0] + dx;
                int ny = player[1] + dy;
                if (nx >= 0 && nx < size && ny >= 0 && ny < size && maze[nx][ny] == 0) {
                    player[0] = nx;
                    player[1] = ny;
                    steps++;
                    // Проверка монет
                    for (Iterator<int[]> it = coins.iterator(); it.hasNext(); ) {
                        int[] c = it.next();
                        if (c[0] == nx && c[1] == ny) {
                            it.remove();
                            collected++;
                            break;
                        }
                    }
                    if (player[0] == end[0] && player[1] == end[1]) {
                        elapsed = (int)((System.currentTimeMillis() - startTime) / 1000);
                        System.out.println(colorize("🎉 Поздравляем! Вы вышли за " + steps + " шагов и " + elapsed + " секунд!", GREEN));
                        System.out.println(colorize("Собрано монет: " + collected, YELLOW));
                        int bonus = collected * 10;
                        System.out.println(colorize("Бонус за монеты: +" + bonus + " очков", YELLOW));
                        int total = steps + bonus;
                        System.out.println(colorize("Итоговый счёт: " + total, BOLD));
                        // Рекорды
                        Record rec = records.get(key);
                        if (rec == null || steps < rec.steps) {
                            Record r = new Record();
                            r.steps = steps;
                            r.time = elapsed;
                            r.coins = collected;
                            records.put(key, r);
                            saveRecords(records);
                            System.out.println(colorize("🏆 Новый рекорд!", GREEN));
                        }
                        System.out.println(colorize("Нажмите любую клавишу для выхода...", YELLOW));
                        System.in.read();
                        return;
                    }
                }
            }
            Thread.sleep(30);
        }
    }
}

File: maze.cs
// maze.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Threading;
using System.Diagnostics;

class MazeGame
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "red" => "\x1b[91m",
            "green" => "\x1b[92m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            "cyan" => "\x1b[96m",
            "gray" => "\x1b[90m",
            "bold" => "\x1b[1m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    const int DEFAULT_SIZE = 10;
    const int DEFAULT_COINS = 5;

    static string RecordFile => Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".maze_records.json");

    static Dictionary<string, Record> LoadRecords()
    {
        try
        {
            string json = File.ReadAllText(RecordFile);
            return JsonSerializer.Deserialize<Dictionary<string, Record>>(json) ?? new Dictionary<string, Record>();
        }
        catch { return new Dictionary<string, Record>(); }
    }

    static void SaveRecords(Dictionary<string, Record> records)
    {
        string json = JsonSerializer.Serialize(records, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(RecordFile, json);
    }

    class Record
    {
        public int Steps { get; set; }
        public int Time { get; set; }
        public int Coins { get; set; }
    }

    static void ClearScreen() => Console.Clear();

    static int[,] GenerateMaze(int size)
    {
        var maze = new int[size, size];
        for (int i = 0; i < size; i++)
            for (int j = 0; j < size; j++)
                maze[i, j] = 1;

        var rand = new Random();
        void Dfs(int x, int y)
        {
            maze[x, y] = 0;
            var dirs = new (int dx, int dy)[] { (0, 2), (0, -2), (2, 0), (-2, 0) };
            for (int i = dirs.Length - 1; i > 0; i--)
            {
                int j = rand.Next(i + 1);
                var tmp = dirs[i];
                dirs[i] = dirs[j];
                dirs[j] = tmp;
            }
            foreach (var d in dirs)
            {
                int nx = x + d.dx, ny = y + d.dy;
                if (nx > 0 && nx < size - 1 && ny > 0 && ny < size - 1 && maze[nx, ny] == 1)
                {
                    maze[x + d.dx / 2, y + d.dy / 2] = 0;
                    Dfs(nx, ny);
                }
            }
        }
        Dfs(1, 1);
        maze[size - 2, size - 1] = 0;
        maze[size - 1, size - 1] = 0;
        return maze;
    }

    static List<(int, int)> PlaceCoins(int[,] maze, int numCoins, int size)
    {
        var coins = new List<(int, int)>();
        var rand = new Random();
        int attempts = 0;
        while (coins.Count < numCoins && attempts < 1000)
        {
            int x = rand.Next(1, size - 1);
            int y = rand.Next(1, size - 1);
            if (maze[x, y] == 0 && !coins.Contains((x, y)) && !(x == 1 && y == 1) && !(x == size - 2 && y == size - 1))
            {
                coins.Add((x, y));
            }
            attempts++;
        }
        return coins;
    }

    static List<(int, int)> FindPath(int[,] maze, (int, int) start, (int, int) end, int size)
    {
        var visited = new bool[size, size];
        var parent = new (int, int)[size, size];
        var queue = new Queue<(int, int)>();
        queue.Enqueue(start);
        visited[start.Item1, start.Item2] = true;
        var dirs = new (int, int)[] { (0, 1), (0, -1), (1, 0), (-1, 0) };
        while (queue.Count > 0)
        {
            var cur = queue.Dequeue();
            if (cur == end)
            {
                var path = new List<(int, int)>();
                var c = cur;
                while (c != start)
                {
                    path.Add(c);
                    c = parent[c.Item1, c.Item2];
                }
                path.Add(start);
                path.Reverse();
                return path;
            }
            foreach (var d in dirs)
            {
                int nx = cur.Item1 + d.Item1, ny = cur.Item2 + d.Item2;
                if (nx >= 0 && nx < size && ny >= 0 && ny < size && !visited[nx, ny] && maze[nx, ny] == 0)
                {
                    visited[nx, ny] = true;
                    parent[nx, ny] = cur;
                    queue.Enqueue((nx, ny));
                }
            }
        }
        return null;
    }

    static void DrawMaze(int[,] maze, (int, int) player, List<(int, int)> coins, int size, bool showPath, List<(int, int)> path)
    {
        ClearScreen();
        Console.WriteLine(Colorize("┌" + new string('─', size * 2 - 1) + "┐", "gray"));
        for (int i = 0; i < size; i++)
        {
            string line = "│";
            for (int j = 0; j < size; j++)
            {
                if (i == player.Item1 && j == player.Item2)
                    line += Colorize("@", "green");
                else if (showPath && path != null && path.Contains((i, j)))
                    line += Colorize(".", "cyan");
                else if (coins.Contains((i, j)))
                    line += Colorize("💎", "yellow");
                else if (maze[i, j] == 1)
                    line += Colorize("█", "gray");
                else
                    line += " ";
                if (j < size - 1) line += " ";
            }
            line += "│";
            Console.WriteLine(line);
        }
        Console.WriteLine(Colorize("└" + new string('─', size * 2 - 1) + "┘", "gray"));
    }

    static int Main(string[] args)
    {
        int size = DEFAULT_SIZE;
        int coinsCount = DEFAULT_COINS;
        bool showPath = false;

        for (int i = 0; i < args.Length; i++)
        {
            if (args[i] == "-s" && i + 1 < args.Length) size = int.Parse(args[++i]);
            else if (args[i] == "-c" && i + 1 < args.Length) coinsCount = int.Parse(args[++i]);
            else if (args[i] == "-p") showPath = true;
            else if (args[i] == "-h")
            {
                Console.WriteLine("Usage: maze [options]\n  -s <N>   Size (default 10)\n  -c <N>   Coins (default 5)\n  -p       Show path");
                return 0;
            }
        }
        if (size % 2 == 0) size++;
        if (size < 5) size = 5;

        var maze = GenerateMaze(size);
        var coins = PlaceCoins(maze, coinsCount, size);
        var start = (1, 1);
        var end = (size - 2, size - 1);
        var player = (1, 1);
        int steps = 0;
        var stopwatch = Stopwatch.StartNew();
        int collected = 0;
        var records = LoadRecords();
        string key = $"{size}x{size}";

        List<(int, int)> path = null;
        if (showPath) path = FindPath(maze, start, end, size);

        Console.CancelKeyPress += (sender, e) =>
        {
            Console.WriteLine(Colorize("\nИгра прервана.", "yellow"));
            Environment.Exit(0);
        };

        while (true)
        {
            DrawMaze(maze, player, coins, size, showPath, path);
            int elapsed = (int)stopwatch.Elapsed.TotalSeconds;
            Console.WriteLine(Colorize($"Шаги: {steps}  Время: {elapsed}с  Монеты: {collected}/{coins.Count}", "bold"));
            Console.WriteLine(Colorize("Управление: стрелки (WASD), Q - выход", "blue"));
            if (showPath && path != null) Console.WriteLine(Colorize("Кратчайший путь показан точками (cyan)", "yellow"));

            var keyInfo = Console.ReadKey(true);
            var keyChar = char.ToLower(keyInfo.KeyChar);
            int dx = 0, dy = 0;
            if (keyChar == 'q')
            {
                Console.WriteLine(Colorize("Выход из игры.", "yellow"));
                break;
            }
            if (keyInfo.Key == ConsoleKey.UpArrow || keyChar == 'w') dx = -1;
            else if (keyInfo.Key == ConsoleKey.DownArrow || keyChar == 's') dx = 1;
            else if (keyInfo.Key == ConsoleKey.LeftArrow || keyChar == 'a') dy = -1;
            else if (keyInfo.Key == ConsoleKey.RightArrow || keyChar == 'd') dy = 1;

            if (dx != 0 || dy != 0)
            {
                int nx = player.Item1 + dx;
                int ny = player.Item2 + dy;
                if (nx >= 0 && nx < size && ny >= 0 && ny < size && maze[nx, ny] == 0)
                {
                    player = (nx, ny);
                    steps++;
                    if (coins.Contains((nx, ny)))
                    {
                        coins.Remove((nx, ny));
                        collected++;
                    }
                    if (player == end)
                    {
                        elapsed = (int)stopwatch.Elapsed.TotalSeconds;
                        Console.WriteLine(Colorize($"🎉 Поздравляем! Вы вышли за {steps} шагов и {elapsed} секунд!", "green"));
                        Console.WriteLine(Colorize($"Собрано монет: {collected}", "yellow"));
                        int bonus = collected * 10;
                        Console.WriteLine(Colorize($"Бонус за монеты: +{bonus} очков", "yellow"));
                        int total = steps + bonus;
                        Console.WriteLine(Colorize($"Итоговый счёт: {total}", "bold"));
                        if (!records.ContainsKey(key) || steps < records[key].Steps)
                        {
                            records[key] = new Record { Steps = steps, Time = elapsed, Coins = collected };
                            SaveRecords(records);
                            Console.WriteLine(Colorize("🏆 Новый рекорд!", "green"));
                        }
                        Console.WriteLine(Colorize("Нажмите любую клавишу для выхода...", "yellow"));
                        Console.ReadKey(true);
                        return 0;
                    }
                }
            }
        }
        return 0;
    }
}
